kythe/cxx/indexer/cxx/IndexerASTHooks.cc

```diff
@@ -3827,856 +3992,606 @@ void IndexerASTVisitor::DumpTypeContext(unsigned Depth, unsigned Index) {
+  // TODO(shahms): Caching.
+  const auto* IFace = CHECK_NOTNULL(TL.getIFaceDecl());
+  // Link to the implementation if we have one, otherwise link to the
+  // interface. If we just have a forward declaration, link to the nominal
+  // type node.
+  if (const auto* Impl = IFace->getImplementation()) {
+    return {BuildNodeIdForDecl(Impl), Claimability::Unclaimable};
+  } else if (!IsObjCForwardDecl(IFace)) {
+    return {BuildNodeIdForDecl(IFace), Claimability::Unclaimable};
+  } else {
+    // Thanks to the ODR, we shouldn't record multiple nominal type nodes
+    // for the same TU: given distinct names, NameIds will be distinct,
+    // there may be only one definition bound to each name, and we
+    // memoize the NodeIds we give to types.
+    return BuildNominalNodeIdForDecl(IFace);
+  }
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForPointer(clang::PointerTypeLoc TL) {
+  if (auto PointeeID = BuildNodeIdForType(TL.getPointeeLoc())) {
+    return ApplyBuiltinTypeConstructor("ptr", *PointeeID);
+  }
+  return NodeSet::Empty();
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForLValueReference(
+    clang::LValueReferenceTypeLoc TL) {
+  if (auto PointeeID = BuildNodeIdForType(TL.getPointeeLoc())) {
+    return ApplyBuiltinTypeConstructor("lvr", *PointeeID);
+  }
+  return NodeSet::Empty();
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForRValueReference(
+    clang::RValueReferenceTypeLoc TL) {
+  if (auto PointeeID = BuildNodeIdForType(TL.getPointeeLoc())) {
+    return ApplyBuiltinTypeConstructor("rvr", *PointeeID);
+  }
+  return NodeSet::Empty();
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForAuto(clang::AutoTypeLoc TL) {
+  return BuildNodeSetForDeduced(TL);
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForDeducedTemplateSpecialization(
+    clang::DeducedTemplateSpecializationTypeLoc TL) {
+  return BuildNodeSetForDeduced(TL);
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForDeduced(clang::DeducedTypeLoc TL) {
+  auto DeducedQT = TL.getTypePtr()->getDeducedType();
+  if (DeducedQT.isNull()) {
+    // We still need to come up with a name here--it's more useful than
+    // returning None, since we might be down a branch of some structural
+    // type. We might also have an unconstrained type variable,
+    // as with `auto foo();` with no definition.
+    // TODO(zarko): Is "auto" the correct thing to return here for
+    // a DeducedTemplateSpecialization?
+    return Observer.getNodeIdForBuiltinType("auto");
+  }
+  return BuildNodeSetForType(DeducedQT);
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForQualified(
+    clang::QualifiedTypeLoc TL) {
+  // TODO(zarko): ObjC tycons; embedded C tycons (address spaces).
+  if (auto ID = BuildNodeIdForType(TL.getUnqualifiedLoc())) {
+    // Don't look down into type aliases. We'll have hit those during the
+    // BuildNodeIdForType call above.
+    // TODO(zarko): also add canonical edges (what do we call the edges?
+    // 'expanded' seems reasonable).
+    //   using ConstInt = const int;
```

Comment:
delete?

Reply:
No, this is still as relevant as it ever was.

Reply:
The commented out using lines?